core/modules/wbase/SendChannelShared.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_WBASE_SENDCHANNELSHARED_H
+#define LSST_QSERV_WBASE_SENDCHANNELSHARED_H
+
+// System headers
+#include <functional>
+#include <memory>
+#include <queue>
+#include <string>
+#include <stdexcept>
+#include <thread>
+
+// Qserv headers
+#include "global/Bug.h"
+#include "wbase/SendChannel.h"
+#include "wbase/TransmitData.h"
+#include "wcontrol/TransmitMgr.h"
+
+namespace lsst {
+namespace qserv {
+namespace wbase {
+
+/// A class that provides a SendChannel object with synchronization so it can be
+/// shared by across multiple threads. Due to what may be sent, the synchronization locking
+/// is needs to be available outside of the class.
+class SendChannelShared {
+public:
+    using Ptr = std::shared_ptr<SendChannelShared>;
+
+    /// To help ensure that _streamMutex is locked before calling,
+    /// many member functions require a StreamGuard argument.
+    using StreamGuard = std::lock_guard<std::mutex> const&;
+
+    SendChannelShared()=delete;
+    SendChannelShared(SendChannelShared const&) = delete;
+    SendChannelShared& operator=(SendChannelShared const&) = delete;
+
+    static Ptr create(SendChannel::Ptr const& sendChannel, wcontrol::TransmitMgr::Ptr const& transmitMgr);
+
+    ~SendChannelShared() = default;
```

Comment:
This is also weird. Prior code made sure the channel was not active should the destructor be called and actually tried to kill it so that bad things wouldn't happen. Now it seems it doesn't matter yet the code hasn't changed from the "active" perspective. What am I missing?

Reply:
I'll take look. Mostly the changes here had to do with stopping a thread that no longer exists but it does seem like the SendChannel should be checked.

Reply:
I put the kill() back, but it doesn't do much besides set a flag. When there was a thread, this would cause the thread to finish quickly without trying to send anything. Since the thread is gone, it doesn't have much effect anymore. SendChannel cleans up itself when it is deleted.